server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java

```diff
@@ -304,52 +344,65 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
-            try (TraceStage stage = sqlTuple.getSqlWatch().start(SqlExecuteStages.EXECUTE)) {
+    protected List<JdbcGeneralResult> doExecuteSql(Statement statement, SqlTuple sqlTuple, CountDownLatch latch) {
+        try {
+            String sql = sqlTuple.getExecutedSql();
+            if (!ifFunctionCallExists(sql)) {
+                // use text protocal
+                try (TraceStage stage = sqlTuple.getSqlWatch().start(SqlExecuteStages.EXECUTE)) {
+                    boolean isResultSet;
+                    try {
+                        isResultSet = statement.execute(sql);
+                    } catch (SQLException e) {
+                        return handleException(e, statement, sqlTuple);
+                    }
+                    latch.countDown();
+                    return consumeStatement(statement, sqlTuple, isResultSet);
+                }
+            }
+            // use ps protocal
+            String preparedSql = OBJECT_VALUE_PATTERN.matcher(sql).replaceAll("?");
+            log.info(
+                    "Load_file call is detected in sql, use ps protocol to rewrite "
+                            + "the original sql, originalSql={}, modifiedSql={}",
+                    sql, preparedSql);
+            List<FunctionDefinition> definitions = retrieveFunctionCalls(sql);
+            log.info("There is a function call in sql, functions={}", definitions);
+            try (PreparedStatement preparedStatement = statement.getConnection().prepareStatement(preparedSql);
+                    TraceStage stage = sqlTuple.getSqlWatch().start(SqlExecuteStages.EXECUTE)) {
+                for (int i = 0; i < definitions.size(); i++) {
+                    FunctionDefinition definition = definitions.get(i);
+                    if ("load_file".equalsIgnoreCase(definition.getFunctionName())) {
+                        // load binary data
+                        String fileName = retrieveFileNameFromParameters(definition);
+                        File file = nullSafeFindFileByName(fileName);
+                        preparedStatement.setBinaryStream(i + 1, new FileInputStream(file));
+                    } else if ("load_clob_file".equalsIgnoreCase(definition.getFunctionName())) {
+                        // load clob data
+                        String fileName = retrieveFileNameFromParameters(definition);
+                        File file = nullSafeFindFileByName(fileName);
+                        preparedStatement.setClob(i + 1, new FileReader(file));
+                    } else {
+                        throw new NotImplementedException("Unsupport function call " + definition.getFunctionName());
+                    }
+                }
                 boolean isResultSet;
                 try {
-                    isResultSet = statement.execute(sql);
-                } catch (Exception e) {
+                    isResultSet = preparedStatement.execute();
+                } catch (SQLException e) {
```

Comment:
what if other exception is thrown in this calling?

Reply:
fixed this.